src/staking_utils.rs

```diff
@@ -0,0 +1,236 @@
+use hashbrown::HashMap;
+use solana_runtime::bank::Bank;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::vote_program::VoteState;
+
+/// Looks through vote accounts, and finds the latest slot that has achieved
+/// supermajority lockout
+pub fn get_supermajority_slot(bank: &Bank, epoch_height: u64) -> Option<u64> {
+    // Find the amount of stake needed for supermajority
+    let stakes_and_lockouts = epoch_stakes_and_lockouts(bank, epoch_height);
+    let total_stake: u64 = stakes_and_lockouts.values().map(|s| s.0).sum();
+    let supermajority_stake = total_stake * 2 / 3;
+
+    // Filter out the states that don't have a max lockout
+    find_supermajority_slot(supermajority_stake, stakes_and_lockouts.values())
+}
+
+pub fn staked_nodes(bank: &Bank) -> HashMap<Pubkey, u64> {
+    staked_nodes_extractor(bank, |stake, _| stake)
+}
+
+/// Return the checkpointed stakes that should be used to generate a leader schedule.
+pub fn staked_nodes_at_epoch(bank: &Bank, epoch_height: u64) -> HashMap<Pubkey, u64> {
+    staked_nodes_at_epoch_extractor(bank, epoch_height, |stake, _| stake)
+}
+
+/// Return the checkpointed stakes that should be used to generate a leader schedule.
+/// state_extractor takes (stake, vote_state) and maps to an output.
+fn staked_nodes_at_epoch_extractor<F, T>(
+    bank: &Bank,
+    epoch_height: u64,
+    state_extractor: F,
+) -> HashMap<Pubkey, T>
+where
+    F: Fn(u64, &VoteState) -> T,
+{
+    let epoch_slot_height = epoch_height * bank.slots_per_epoch();
+    staked_nodes_at_slot_extractor(bank, epoch_slot_height, state_extractor)
+}
+
+/// Return the checkpointed stakes that should be used to generate a leader schedule.
+/// state_extractor takes (stake, vote_state) and maps to an output
+fn staked_nodes_at_slot_extractor<F, T>(
+    bank: &Bank,
+    current_slot_height: u64,
+    state_extractor: F,
+) -> HashMap<Pubkey, T>
+where
+    F: Fn(u64, &VoteState) -> T,
+{
+    let slot_height = current_slot_height.saturating_sub(bank.stakers_slot_offset());
+
+    let parents = bank.parents();
+    let mut banks = vec![bank];
+    banks.extend(parents.iter().map(|x| x.as_ref()));
+
+    let bank = banks
+        .iter()
+        .find(|bank| bank.slot_height() <= slot_height)
```

Comment:
prefer you use bank.id() over bank.slot_height().  slot_height() is calculated from tick_height, which is often incorrect for a fresh bank.  also, we only want frozen banks, but I think we're getting that by accident?

I'm not sure I'm comfortable with an approach that relies on other parts of the system not squash()ing.  What would make me more comfortable is to have stake should be calculated as we freeze(), and cached somewhere outside the Bank.